### DISCUSSION TITLE
Digit Rollover - Wrong value detection - Model

### DISCUSSION THREAD

**Discussion by gerdur** [jomjol/AI-on-the-edge-device#1318] [General]
Hello, unfortunately I didn't find a clear answer to my following topic:
For a quite long period, I get an Error due to a "Neg Rate Read" when there is a digit rollover.
![Water1](https://user-images.githubusercontent.com/44534489/201108429-c0bce88c-984f-4140-af46-63adbb8e6ce3.png)
For me, the number recognition looks plausible:
![Water2](https://user-images.githubusercontent.com/44534489/201108900-18235fa6-6f48-44b1-84cd-3dc73793c0cf.png)
Based on the Wiki-Documentation I have chosen the dig-class100-0130_s2_q.tflite model
![Water3](https://user-images.githubusercontent.com/44534489/201109467-4551267d-0583-4743-840a-2e84bfc47317.png)
Can you please give me some hint: 
What do I have to change, to get plausible valid values from my watermeter, even if a rollover on the last digits happens? 
Do I have to change the model or some other configuration settings?
Many thanks for your answer!

**Comment by jomjol** [comment]
Unfortunatly there is nothing you can do about this currently. See here: https://github.com/jomjol/AI-on-the-edge-device/wiki/FAQs#even-after-i-have-setup-everything-perfect-there-is-a-false-reading---especially-around-the-zero-crossing-roll-over-to-next-number

**Reply by gerdur** [reply]
Hi jomjol, many thanks for your feedback. At first, this problem doesn't look so complicated but when taking a little bit more time and diving deeper into this topic, it is obviously really difficult to implement (especially to make it work not only for one specific watermeter but a collection of many different units). 
Please allow me one more question: are there any plans to work on this topic and do we have a chance to get a (maybe not 100%) solution in the near future? 
I have to analyze it in more detail, but actually it looks like there is large percentage of invalid values, making it hard to use the results.

**Reply by jomjol** [reply]
> Hi jomjol, many thanks for your feedback. At first, this problem doesn't look so complicated but when taking a little bit more time and diving deeper into this topic, it is obviously really difficult to implement (especially to make it work not only for one specific watermeter but a collection of many different units). Please allow me one more question: are there any plans to work on this topic and do we have a chance to get a (maybe not 100%) solution in the near future? I have to analyze it in more detail, but actually it looks like there is large percentage of invalid values, making it hard to use the results.

Currently my programming capacities are very limited and I'm not working activly on this issue. In the next future (this year) I don't see a change to invest this problem.

**Reply by gerdur** [reply]
Hi jomjol, just one additional thought about this topic:
Is it correct, that the analog readings are nearly 100% reliable and there is only a problem regarding rollover on the digital readings? 
In this case, if I would analyze the values regularly (e.g. every 5 minutes), it would be sufficient to take only the analog readings into account?  Ok, as long as I will not take more than 1000 liters in less than 5 minutes. I think this is true for a normal household. 
So the basic idea is, to mainly use the analog readings for the readout value. This would give me of course a relative and not an absolute value. The absolute readout value would be calculated/synchronized from time to time, by taking the digital value into account, if it is valid. This is just a quick idea and maybe this could be an option to activate in the watermeter settings... 
Best regards, Gerhard (sorry for my confusing explanation :-))

**Reply by jomjol** [reply]
The analog readings is more stable, but I would not call it 100% accurate. The only thing is, that you realize a misreading not so clearly (imagine a missreading +/-5 liter after a minute it would be okay again. But you can rely on the analog readings and calculate the digital by yourself - acutally the first versions did something similar.

**Reply by gerdur** [reply]
Ok, I suppose an appropriate calculation could also be done on Home Assistant level in a lambda function (I will check this). But as a first step it would be of interest to log all analog and ditigal readings to get a feeling, when the problem is starting/ending during a digit rollover. Especially when multiple digits are rolling. Many thanks for your good work.

**Comment by AseKarlsson** [comment]
Hi,

Have the same kind of issues.
Are here Home Assistant users, can the sensor in configuration.yaml (MQTT) be configured to not pass thru negative values?

Thanks,
Ase

**Reply by AseKarlsson** [reply]
I have the same kind of meter and use the following tflite files, are these the recommended ones?
Jomjol, your input is very much appreciated!


![image](https://user-images.githubusercontent.com/80900477/201403275-5f043eb3-c384-4379-b744-8ab406a6f64f.png)

**Reply by jomjol** [reply]
No, this are "old" style files. For more recent files see:
- explanation: https://github.com/jomjol/AI-on-the-edge-device/wiki/Choosing-the-Model
- files: https://github.com/jomjol/AI-on-the-edge-device/tree/rolling/sd-card/config

**Reply by AseKarlsson** [reply]
Thanks Jomjol for your quick reply,
For our kind of meter, what files to select, both digital and analog?
The Wiki I have already checked, but really did not get wiser.

**Reply by jomjol** [reply]
The `dig` files for the digits and the `ana` file for the pointer like readings. Which one exaclty you need to test.

**Reply by AseKarlsson** [reply]
Hi Jomjol,

At the moment I have all these files below on the SD-Card.
If I change my question. Of these files which are obsolete (old) and should not be used anymore?
What exact files do you recommend to be used with our kind of meter, both digigital and analog?

![image](https://user-images.githubusercontent.com/80900477/201468120-fb6239bb-8498-4225-bb23-833ac88a720d.png)
![image](https://user-images.githubusercontent.com/80900477/201468124-f14d44d5-1f8b-4e62-96c1-2a2fc3325c4d.png)

**Reply by jomjol** [reply]
digits: `dig-class11_1440_s2.tflite`
analog: `ana-cont_11.3.0_s2.tflite`

**Reply by caco3** [reply]
> Are here Home Assistant users, can the sensor in configuration.yaml (MQTT) be configured to not pass thru negative values?

You could add a template sensor which can do that.

**Reply by AseKarlsson** [reply]
Thanks!
